python_modules/libraries/dagster-aws/dagster_aws/ecs/launcher.py

```diff
@@ -396,22 +401,59 @@ def launch_run(self, context: LaunchRunContext) -> None:
         if "launchType" in run_task_kwargs and run_task_kwargs.get("capacityProviderStrategy"):
-            del run_task_kwargs["launchType"]
-
-        # Run a task using the same network configuration as this processes's task.
-        response = self.ecs.run_task(**run_task_kwargs)
+            del run_task_kwargs["launchType"]            
+
+        # function to reduce duplicate code and reduce nesting
+        def run_task():
+            response = self.ecs.run_task(**run_task_kwargs)
+            tasks = response["tasks"]
+            if not tasks:
+                failures = response["failures"]
+                exceptions = []
+                for failure in failures:
+                    arn = failure.get("arn")
+                    reason = failure.get("reason")
+                    detail = failure.get("detail")
+                    exceptions.append(Exception(f"Task {arn} failed because {reason}: {detail}"))
+                    if reason == "Capacity is unavailable at this time. Please try again later or in a different availability zone":
+                        raise SpotInstanceException()
+                # Raise regular error if error not due to spot integration
+                raise Exception(exceptions)
+            else:
+                return tasks
 
-        tasks = response["tasks"]
+        # FARGATE_SPOT safety feature 
+        if run_task_kwargs.get("capacityProviderStrategy") == [{'capacityProvider': 'FARGATE_SPOT'}]:
+            # if capacityProviderStrategy is FARGATE_SPOT 
+            #   try to execute 'spot_retries' times to use spot instance 
+            #   else switch to on demand instance
 
-        if not tasks:
-            failures = response["failures"]
-            exceptions = []
-            for failure in failures:
-                arn = failure.get("arn")
-                reason = failure.get("reason")
-                detail = failure.get("detail")
-                exceptions.append(Exception(f"Task {arn} failed because {reason}: {detail}"))
-            raise Exception(exceptions)
+            spot_retries = run.tags.get("ecs/spot_retries", container_context.run_resources.get("spot_retries"))
+            if spot_retries:
+                try:
+                    # ensure value is an integer, else raise error
+                    spot_retries = int(spot_retries)
+                except Exception as e:
+                    raise e
+            else:
+                spot_retries = 0
+
+            for i in range(spot_retries + 2):  # if spot_retries = 0, first try spot [i = 0] and then use on demand [i = 1]
+                if i < (spot_retries + 1):  # if spot_retries = 0, first evaluation is True [i=0 < 1] and second evaluation is False [i=1 !< 1]
+                    try:
+                        tasks = run_task()
+                        break
+                    except SpotInstanceException:
+                        asyncio.sleep(5)
```

Comment:
the interval here should be something configurable as well - you probably would want a configuration dictionary just for capacity retries with the number of retries and the interval between retries